### PR TITLE
feat(convert): export ENV vars

### PR DESCRIPTION
**Pull request by rchincha**
We could have passed these into the "build_env:" directive, however it appears environment substitution is also allowed such as PATH=/usr/local/sbin:$PATH for that layer, which is harder to figure out. Instead just "export" them in the "run:" directive.

Signed-off-by: Ramkumar Chinchani <rchincha@cisco.com>

**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/convert.bats [hunk 1]
```diff
@@ -64,3 +64,17 @@ EOF
   rm -f stacker.yaml stacker-subs.yaml
   stacker clean
 }
+
+@test "build python image" {
+  git clone https://github.com/docker-library/python.git
+  chmod -R a+rwx python
```
Comment (smoser) [inline test/convert.bats:70]: this seems questionable.
you really want to make *everything* under there executable?

File: test/convert.bats [hunk 2]
```diff
@@ -64,3 +64,17 @@ EOF
   rm -f stacker.yaml stacker-subs.yaml
   stacker clean
 }
+
+@test "build python image" {
+  git clone https://github.com/docker-library/python.git
```
Comment (smoser) [inline test/convert.bats:69]: Some day you'll wake up and this repository will have changed and tests will fail.  Is that desirable?
The alternative is to put the test you need here in the repository, which will be more reliable and faster, no?

File: test/convert.bats [hunk 3]
```diff
@@ -64,3 +64,17 @@ EOF
...
+  chmod -R a+rwx python
+  cd python/3.11/alpine3.17
+  stacker convert --docker-file Dockerfile --output-file stacker.yaml --substitute-file stacker-subs.yaml
+  stacker build -f stacker.yaml --substitute-file stacker-subs.yaml --substitute IMAGE=python
+  if [ -z "${REGISTRY_URL}" ]; then
+      skip "skipping test because no registry found in REGISTRY_URL env variable"
+  fi
+  stacker publish -f stacker.yaml --substitute-file stacker-subs.yaml --substitute IMAGE=python --skip-tls --url docker://${REGISTRY_URL} --tag latest
```
Comment (smoser) [inline test/convert.bats:77]: My fault for not paying attention, but it seems like 'REGISTRY_URL' should be a URL, not a hostname/path (I'm assuming that a valid setting would be 'docker.io/library', which isn't a url at all).

**Review by smoser — COMMENTED** [review]
This is an interesting change.
If I understand it correctly, it is so that you can support the following line from [here](https://github.com/docker-library/python/blob/master/3.11/alpine3.17/Dockerfile#L10):

    ENV PATH /usr/local/bin:$PATH

What is most surprising about it is that it seems to me that the above line is almost certainly *wrong*.  If it is setting the PATH environment to be used inside the container, then why would you consult the *hosts* value of PATH?  It should just explicitly set the value for the container like:

    ENV PATH usr/local/bin:/usr/sbin:/usr/bin

If my host had a PATH variable of `/bin` or perhaps '/opt/bin:/opt/sbin', that could be perfectly fine for my host.  But passing that variable through at best makes the execution less declarative and at worst breaks the build.

Maybe I'm missing something?

**Comment by rchincha** [comment]
> This is an interesting change. If I understand it correctly, it is so that you can support the following line from [here](https://github.com/docker-library/python/blob/master/3.11/alpine3.17/Dockerfile#L10):
> 
> ```
> ENV PATH /usr/local/bin:$PATH
> ```
> 
> What is most surprising about it is that it seems to me that the above line is almost certainly _wrong_. If it is setting the PATH environment to be used inside the container, then why would you consult the _hosts_ value of PATH? It should just explicitly set the value for the container like:
> 
> ```
> ENV PATH usr/local/bin:/usr/sbin:/usr/bin
> ```
> 
> If my host had a PATH variable of `/bin` or perhaps '/opt/bin:/opt/sbin', that could be perfectly fine for my host. But passing that variable through at best makes the execution less declarative and at worst breaks the build.
> 
> Maybe I'm missing something?

The $PATH is derived from the FROM section layer. So how does one figure that out?

File: test/convert.bats (hunk 2 above)
Comment (rchincha) [reply test/convert.bats:69]: This was to make sure the **exact** upstream tests work.

File: test/convert.bats (hunk 3 above)
Comment (rchincha) [reply test/convert.bats:77]: This was borrowed from test/publish.bats

File: test/convert.bats (hunk 2 above)
Comment (rchamarthy) [reply test/convert.bats:69]: Is it better to fork and then use the fork?

File: test/convert.bats (hunk 3 above)
Comment (rchincha) [reply test/convert.bats:77]: .github/workflows/build.yaml
49:          REGISTRY_URL: localhost:5000

File: test/convert.bats (hunk 1 above)
Comment (rchincha) [reply test/convert.bats]: fixed

Comment (rchincha) [reply test/convert.bats]: mkdir /home/runner/work/stacker/stacker/stackertest-test_build_python_image.2Okow7/python/3.11/alpine3.17/roots: permission denied couldn't make rootfs dir for testsuite

^ but this. Fixing.

File: test/convert.bats (hunk 3 above)
Comment (smoser) [reply test/convert.bats:77]: yeah, we should fix that to be an actual URL, then it would or could support REGISTRY_URL=oci:mirrors/docker.io/library .  I realize you didn't change it here.

File: test/convert.bats (hunk 2 above)
Comment (smoser) [reply test/convert.bats:69]: I can understand the desire to make sure that "real world examples" work.  I personally hate it when the internet breaks my c-i.  It means someone has to spend time figuring out why the light went from green to red, or worse, people accept "sometimes c-i breaks".

If you want to have a collection of "moving target real world docker files", maybe that is best to do in an integration test of some sort.  Unfortunately that just kicks the "figure out what failed" problem down the line.

It is entirely possible that a stacker change regresses this behavior and this test catches that.  But I think it is more likely that a breakage occurs due to the added dependency on a moving third party.

I would much prefer a quicker that did something like:

```
@test "check environment variable substitution" {
   cat > dockerfile <<EOF
FROM alpine:3.17

ENV PATH /usr/local/bin:$PATH

RUN [ "$PATH" = "/usr/local/bin:/usr/bin:/usr/sbin" ]
EOF
}
```

I like the REGISTRY_URL so that these tests don't have to depend on docker.io.

File: test/convert.bats (hunk 1 above)
Comment (smoser) [reply test/convert.bats]: is that an indication of a larger problem with permissions? It seems unlikely that the problem here is with that specific directory and not more generic.

stacker should only need write permission for the uid owner of that directory in order to create 'roots' (and .stacker and oci) in the directory. 

what was the ownership of alpine3.17/ ?  How does content for the test get copied into that temp dir? maybe umask got foobar'ed somehow?

Comment (rchincha) [reply test/convert.bats]: Let's find out.

File: test/convert.bats (hunk 2 above)
Comment (rchincha) [reply test/convert.bats:69]: That Dockerfile in that repo unlikely to change for a long while and if it breaks or more likely when it disappears, we will look at it and fix.

**Comment by rchincha** [comment]
> That makes me feel much better, that it isn't just taking the current environment's value.

Yup.

File: test/convert.bats (hunk 2 above)
Comment (smoser) [reply test/convert.bats:69]: > That Dockerfile in that repo unlikely to change for a long while and if it breaks or more likely when it disappears, we will look at it and fix.

I disagree with that statement.  It has [changed 5 times in 2 weeks](https://github.com/docker-library/python/commits/0a56f04a07495e0ed61663c73bc0c2e175773ddc/3.11/alpine3.17/Dockerfile) including a revert.

I really think it is very high cost in build time (we're building *python* as part of stacker c-i?) and reliability.  Successful c-i now depends on:
 * access to 'alpine:3.17' that will change over time
 * 'git clone' from github (which does not auto-retry on failure) and basically cannot be mirrored

And all of this is to check the expansion of a templated string.

File: test/convert.bats (hunk 1 above)
Comment (rchincha) [reply test/convert.bats]:
```
# Cloning into 'python'...
# python/:
# total 104
# drwxr-xr-x 10 root   root   4096 Feb 14 17:37 .
# drwx------  5 runner runner 4096 Feb 14 17:37 ..
# drwxr-xr-x  8 root   root   4096 Feb 14 17:37 .git
# -rw-r--r--  1 root   root     94 Feb 14 17:37 .gitattributes
# drwxr-xr-x  3 root   root   4096 Feb 14 17:37 .github
# -rw-r--r--  1 root   root     17 Feb 14 17:37 .gitignore
```

Everything is root-owned, that's why

**Comment by rchincha** [comment]
Let's split this into two parts. A nightly "convert/build" job that works on real-world examples, while we do basic tests under bats.

File: test/convert.bats (hunk 2 above)
Comment (rchincha) [reply test/convert.bats]: @smoser, moved the conversion out of the core ci path, and add a "nightly" conversion workflow, at least some place we can catch issues.

File: pkg/stacker/convert.go [hunk 4]
```diff
@@ -123,7 +129,7 @@ func (c *Converter) convertCommand(cmd *Command) error {
 
 		for _, line := range cmd.Value {
 			// patch some cmds
-			line = strings.ReplaceAll(line, "mkdir", "mkdir -p ")
+			line = strings.ReplaceAll(line, "mkdir", "mkdir -p")
```
Comment (hallyn) [inline pkg/stacker/convert.go:132]: So if the dockerfile was doing

```
alias mkdir_p="mkdir -p $*"
mkdir_p /tmp/a
```
This will trip up.  Do we care?

Why is this needed - does a dockerfile actually do this?  If so, maybe doing a regexp replace for "\<mkdir\>" would at least be safer.

Comment (rchincha) [reply pkg/stacker/convert.go:132]: Yes, we do need it to be compatible with docker behavior. But you also have a point.

Comment (rchincha) [reply pkg/stacker/convert.go:132]: Simplest change maybe to 's/mkdir /mkdir -p /g'

Comment (hallyn) [reply pkg/stacker/convert.go:132]: I was thinking something like

```
re := regexp.MustCompile('\\bmkdir\\b')
line  = re.ReplaceAllString(line, 'mkdir -p')

Comment (rchincha) [reply pkg/stacker/convert.go]: Ok, your regexp suggestion/change went through

**Review by hallyn — APPROVED** [review]

File: pkg/stacker/convert.go (hunk 4 above)
Comment (hallyn) [reply pkg/stacker/convert.go]: cool, thanks.
